packages/model-viewer/package.json

```diff
@@ -73,8 +73,9 @@
     "3d"
   ],
   "dependencies": {
-    "three": "^0.146.0",
-    "lit": "^2.2.3"
+    "lit": "^2.2.3",
+    "postprocessing": "^6.29.1",
```

Comment:
How big is this dependency? What's the before/after on our gzipped bundle?

Reply:
I'll check and get back to you on that

Reply:
This is the main thing I need an answer on - if this has significant size, then it should be optionally, lazily loaded like `lottie-web`. Also note that `three` has been switched to a peer dependency.

Reply:
According to [bundlephobia](https://bundlephobia.com/package/postprocessing@6.30.1) it's twice the size of three.js (presumably three.js is part of this bundled value). This means we definitely need a way for it to be optionally loaded. Figuring that out will be the main API-level design challenge.

Reply:
I am actually seeing very different numbers there. Bundlephobia says, it's pretty much exactly half of [three's](https://bundlephobia.com/package/three@0.150.1) size. Am I looking at it wrong?

Reply:
postprocessing doesn't bundle three.js AFAIK, but it's a big library of effects and shaders, so tree-shaking is going to be critical. If the build system here is configured for it, tree-shaking should work well.

Reply:
![image](https://user-images.githubusercontent.com/54665621/223191289-e36f2f06-14c7-4a32-a379-825a20abdab1.png)
in vs-code i see that the bundle size according to [Import Cost ](https://marketplace.visualstudio.com/items?itemName=wix.vscode-import-cost) is ~100kb, compared to ~400kb for three.js. 
Regardless, as @donmccurdy said if tree-shaking is supported I don't believe we need to import all of the possible effects and shaders, so that could definitely be brought down. Lazy loading probably will require additional checks because the postprocessing sits very much in the Renderer component with three.js, I'm not sure thats an optimal solution.

Reply:
Yes, we use rollup for our builds, which tree shakes. So: can I have a before/after of our gzipped bundle size? If it's too much, then it may be better to make a `<model-viewer++>` component that depends on both. I'm open to options, but I'd like to see some numbers in order to make a decision.

Reply:
@blumelume Ah yes, I read it wrong, so only 2/3 the size (gzipped). Still, that's far too big, so let's see where tree-shaking gets us. Our bundle is 210kb all up (gzipped), though I just realized that bundlephobia now shows us as 55kb due to three.js now being a peer dependency. 🙄 I would say bloating that by 10% (20kb) would definitely be too much unless we had a lot of users of this feature.

Reply:
![image](https://user-images.githubusercontent.com/54665621/223208627-80b83213-d99d-4bbf-92ee-3f3a77680590.png)

<html>
<body>
<!--StartFragment-->

build/postprocessing.esm.js | 102.33KB | 11.5%
-- | -- | --


<!--EndFragment-->
</body>
</html>

note that I am still on the older version (@2.1.1) as I haven't merged the latest changes from master which make three into a peer-dependency, so the size without the postprocessing is ~800kb).

Reply:
postprocessing is tree-shakeable according to bundlephobia

Reply:
Thank you @Beilinson! Okay, so between this increase and the antialiasing difference above, this feels like it should be somehow optional, since this won't be used by most of our users (and certainly there is performance impact, particularly on mobile). I'm not sure the best path forward - a simple option from my perspective is you can simply fork our repo into `<model-viewer-postprocessing>` (or something) and make this advanced version. I'll be happy to point advanced users your way. If we can find a decent way to make this optional in the repo here, that would likely be preferable, but I'm not quite sure how. @donmccurdy any thoughts here?

Reply:
I do like the idea of child components, and I'm considering a similar pattern for a multi-model API. @donmccurdy Can you elaborate on why bundle-splitting and CDN distribution is unappealing (I'm a bit ignorant in this domain)? 

@Beilinson Does this sound like a viable path forward to you?

Reply:
By bundle-splitting, I mean having your build process create multiple JavaScript files for a single npm package, and dynamically importing parts as needed. Great idea in theory but in practice there are a lot of obstacles with build tools, CDN caching, etc. Works well for *application* authors, but I haven't seen library authors doing this successfully before, and wouldn't be excited to learn why the hard way.

Unless the tooling situation improves and there are good examples to follow, I think it's better and safer to just have discrete packages, or discrete exports, and let the user explicitly import what they need. I think the child component approach would fit well with that design.

Reply:
I also like the idea of composition, especially as it solves the issues of new effects and maintenance as well as allows for simple definition of how to expose the effect api, each effect is just a web component and as much as the entire settings interface could be exposed.

I am wondering if its already doable using the <code>externalRenderer</code> on the ModelScene, as that is already called using the internal render loop in the <code><Renderer></code>.  From my testing running both render functions of the effectComposer and the threeRenderer still works, though its a shame from a performance standpoint to run them both. 

I can create a quick mockup using a similar idea to the externalRenderer, i.e <code><postprocessingRenderer></code> or something along those lines, so that the user created externalRenderer will not be overridden by the postprocessing, as well as a basic web component effect or two. 

@donmccurdy @elalish Do you think its best to keep the <code><model-viewer-effects></code> as a separate package then, or just be additional exports by <code><model-viewer></code>? With the second option I believe tree-shaking would work only if the package was imported using JS, with regular html script tag it would probably import everything and that brings us back to square one.

Reply:
I would not use additional ESM exports, for the reason you mention. Using additional [package.json#exports entrypoints](https://nodejs.org/api/packages.html#package-entry-points) would be the way to go if @elalish would like for this to be part of the same npm package, so there are multiple scripts on the CDN, but just one npm package. Otherwise effects could be a separate package, either within this repository or elsewhere.

Reply:
I would avoid `externalRenderer` - it voids much of the useful code we have for managing multiple MV elements performantly and such. That's more for something like cloud streaming. 

We may simply need to expose our internal `Renderer` class (or part of it) to a post-processing web component. And we may want to rely on some static configuration to take care of e.g. setting antialiasing (since it needs to happen before the `Renderer` is created). 

Another advantage to making a more generic plug-in model is we may also be able to incorporate effects like these: https://github.com/0beqz/realism-effects

Reply:
this might be of interest https://webgi.xyz/docs/intro
a threejs based solution with a quite exhaustive post-processing pipeline the editor is worth cheking out

Reply:
![image](https://user-images.githubusercontent.com/54665621/223281374-a2e4e8da-564a-4ffc-874c-bb9547f4362e.png)

I just tested the basic composition idea, it works relatively well out of the gate, will need some work but its definitely the way to go.